File: go/tests/cache_concurrency_test.go
package tests

// TC9 — Cache thread-safety. The TTL cache (cacheStore) is shared mutable state
// on the client; the caching suite only exercises it single-threaded. These fire
// concurrent reads and concurrent mutations at one cached client and assert no
// corruption/panic and a consistent result.
//
// CRITICAL: run this suite under the race detector — an unsynchronized cache map
// would be a real data race:
//
//	go test -race -run TestCacheConcurrency ./...
//
// Parity with csharp/tests/TestCacheConcurrency.cs.

import (
	"context"
	"sync"
	"testing"

	airtable "myairtabletests/output"
)

func TestCacheConcurrency(t *testing.T) {
	requireCreds(t)
	ctx := context.Background()

	t.Run("ConcurrentGetsOfTheSameRecordAreConsistent", func(t *testing.T) {
		at := cachedAirtable(t, 60)
		key := primaryKey("CacheConc", "Reads")
		created, err := at.Primary.CreateOne(ctx, &airtable.PrimaryModel{
			PrimaryKey: airtable.String(key),
		})
		if err != nil {
			t.Fatalf("create: %v", err)
		}
		id := created.ID()
		defer bestEffortDelete(at, id)

		// 25 concurrent gets race to populate/read the shared cache.
		const n = 25
		var wg sync.WaitGroup
		results := make([]*airtable.PrimaryModel, n)
		errs := make([]error, n)
		for i := 0; i < n; i++ {
			wg.Add(1)
			go func(idx int) {
				defer wg.Done()
				results[idx], errs[idx] = at.Primary.GetOne(ctx, id)
			}(i)
		}
		wg.Wait()

		for i := 0; i < n; i++ {
			if errs[i] != nil {
				t.Fatalf("concurrent get %d: %v", i, errs[i])
			}
			m := results[i]
			if m == nil {
				t.Fatalf("concurrent get %d returned nil model", i)
			}
			if m.PrimaryKey == nil || *m.PrimaryKey != key {
				t.Fatalf("concurrent get %d PrimaryKey = %v, want %q", i, m.PrimaryKey, key)
			}
			if m.ID() != id {
				t.Fatalf("concurrent get %d id = %s, want %s", i, m.ID(), id)
			}
		}
	})

	t.Run("ConcurrentReadsAndMutationsDoNotCorruptTheCache", func(t *testing.T) {
		at := cachedAirtable(t, 60)
		key := primaryKey("CacheConc", "Mix")
		created, err := at.Primary.CreateOne(ctx, &airtable.PrimaryModel{
			PrimaryKey: airtable.String(key),
		})
		if err != nil {
			t.Fatalf("create: %v", err)
		}
		id := created.ID()
		defer bestEffortDelete(at, id)

		tableID := (&airtable.PrimaryModel{}).TableID()

		// Interleave reads (populate cache), per-table invalidations, and
		// invalidate-all concurrently ~15x.
		const rounds = 15
		var wg sync.WaitGroup
		errs := make([]error, rounds)
		for i := 0; i < rounds; i++ {
			wg.Add(3)
			go func(idx int) {
				defer wg.Done()
				_, errs[idx] = at.Primary.GetOne(ctx, id)
			}(i)
			go func() {
				defer wg.Done()
				at.Client().InvalidateTableCache(tableID)
			}()
			go func() {
				defer wg.Done()
				at.InvalidateAllCaches()
			}()
		}
		wg.Wait()

		for i, e := range errs {
			if e != nil {
				t.Fatalf("interleaved get %d: %v", i, e)
			}
		}

		// The client is still usable and returns the correct value afterward.
		fetched, err := at.Primary.GetOne(ctx, id)
		if err != nil {
			t.Fatalf("final get: %v", err)
		}
		if fetched.PrimaryKey == nil || *fetched.PrimaryKey != key {
			t.Fatalf("final PrimaryKey = %v, want %q", fetched.PrimaryKey, key)
		}
	})
}

File: javascript/tests/cache-concurrency.test.js
import { describe, it, expect } from "vitest";
const { Record: ATRecord } = require("airtable");
const { Airtable } = require("../output");

/**
 * TC9 — Cache thread-safety.
 *
 * Mirrors csharp/tests/TestCacheConcurrency.cs. The TTL cache is shared mutable state on the
 * client; the caching suite only exercises it single-threaded. JavaScript is single-threaded but
 * async-concurrent: many in-flight promises can interleave around the cache between awaits. These
 * tests fire concurrent reads and concurrent reads+invalidations at one cached client and assert
 * no corruption and a consistent result afterward.
 *
 * Per-table invalidation -> airtable.primary.invalidateCache(); invalidate-all (cascade across all
 * tables) -> airtable.invalidateCache().
 */

function cached() {
	return new Airtable({ cacheSeconds: 60 });
}

function newPrimaryRecord(airtable) {
	return new ATRecord(airtable.primary._table, "", {});
}

function primaryKey(label) {
	return `CacheConcJS ${label} ${Date.now()}-${Math.floor(Math.random() * 1e6)}`;
}

async function bestEffortDelete(airtable, id) {
	if (!id) return;
	try {
		await airtable.primary.delete(id);
	} catch {
		// best-effort cleanup
	}
}

describe("Concurrent gets of the same record are consistent", async () => {
	const airtable = cached();
	const key = primaryKey("Reads");
	const newRecord = newPrimaryRecord(airtable);
	newRecord.set("Primary Key", key);
	const created = await airtable.primary.create(newRecord);
	const id = created.id;

	it("25 concurrent gets all return the correct primaryKey + id", async () => {
		// 25 concurrent gets race to populate/read the shared cache.
		const results = await Promise.all(Array.from({ length: 25 }, () => airtable.primary.get(id)));
		expect(results).toHaveLength(25);
		for (const m of results) {
			expect(m.primaryKey).toBe(key);
			expect(m.id).toBe(id);
		}
	}, 30000);

	it("cleanup", async () => {
		await bestEffortDelete(airtable, id);
	});
});

describe("Concurrent reads and invalidations do not corrupt the cache", async () => {
	const airtable = cached();
	const key = primaryKey("Mix");
	const newRecord = newPrimaryRecord(airtable);
	newRecord.set("Primary Key", key);
	const created = await airtable.primary.create(newRecord);
	const id = created.id;

	it("interleaved reads + invalidations leave the client usable and correct", async () => {
		// Interleave reads (populate cache), per-table invalidations, and invalidate-all
		// (cascade) concurrently.
		const ops = [];
		for (let i = 0; i < 15; i++) {
			ops.push(airtable.primary.get(id));
			ops.push(Promise.resolve().then(() => airtable.primary.invalidateCache()));
			ops.push(Promise.resolve().then(() => airtable.invalidateCache()));
		}
		await Promise.all(ops);

		// The client is still usable and returns the correct value afterward.
		const fetched = await airtable.primary.get(id);
		expect(fetched.primaryKey).toBe(key);
		expect(fetched.id).toBe(id);
	}, 30000);

	it("cleanup", async () => {
		await bestEffortDelete(airtable, id);
	});
});
